### PR TITLE
Add OCS API for sharees list

### LINKED ISSUE #18117: [OCS-API] get sharing recipients
Followup on https://github.com/owncloud/core/issues/16646

Mobile, desktop and web clients need to query the server for sharing recipients (Sharees).

This is the URL to be used: 
http://localhost/owncloud/ocs/v1.php/apps/files_sharing/api/v1/sharees

Allowed method: GET

Query strings:

Key | type | description | default
------------ | ------------- | ------ | -----
search | string | the search string |
itemType | string | the type which is shared (e.g. file or folder) |
shareType[] | integer | any of the shareTypes (0,1, 4, 6) | 
page | integer | the page number to be returned | 1
perPage | integer | the number of items per page | 200

Result:
For each sharee there will be on entry in this form:
```json
{
  "label": "admin",
  "value": {
    "shareType": 1,
    "shareWith": "admin"
  }
}
```

Label is the human readable display name.
shareType identifies the type (user, group, email, remote)
shareWith is the identifier to be use on the sharing api call

Pagination shall be supported as per https://developer.github.com/v3/#pagination

### PR DESCRIPTION
**Pull request by nickvergessen**
@DeepDiver1975 @rullzer 

Fix #18117
### Todo
- [x] Add a filter that prevents sharing with the same user/group/remote again
- [x] Add tests for all methods
- [x] Make shareType an array
- [x] Turn existingShares into an array of share IDs instead of strings
### Related/Depending
- [x] https://github.com/owncloud/core/pull/18352 Autocomplete blocker
- [ ] https://github.com/owncloud/core/pull/18236 Change share dropdown request

### CODE REVIEW

**Comment by nickvergessen** [comment]
code_coverage_for_home_nickv_owncloud_master_core_apps_files_sharing_api_sharees php_-_2015-08-12_15 09 23

**Comment by DeepDiver1975** [comment]
As per specification multiple shareTypes can be passed in

```sh
$ curl -g -v -u admin "http://localhost:8080/ocs/v1.php/apps/files_sharing/api/v1/sharees?itemType=file&shareType[]=0&shareType[]=1"

File: apps/files_sharing/api/sharees.php [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+	 * @param array $potentialSharees
+	 * @param array $existingSharees
+	 * @return array
+	 */
+	protected function filterSharees($potentialSharees, $existingSharees) {
+		$sharees = array_map(function ($sharee) use ($existingSharees) {
+			return in_array($sharee['value']['shareWith'], $existingSharees) ? null : $sharee;
+		}, $potentialSharees);
+
+		return array_filter($sharees);
```
Comment (rullzer) [inline apps/files_sharing/api/sharees.php]: why do the array_map first and then the array_filter. Just 1 array_filter with callback should do the trick

File: apps/files_sharing/api/sharees.php [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+				'itemType' => $itemType,
+				'existingShares' => $existingShares,
+				'page' => $page + 1,
+				'limit' => $perPage,
+			];
+			if ($shareType !== null) {
+				$params['shareType'] = $shareType;
+			}
+			$url = $this->urlGenerator->getAbsoluteURL('/ocs/v1.php/apps/files_sharing/api/v1/sharees') . '?' . http_build_query($params);
+			$response->addHeader('Link', '<' . $url . '> rel="next"');
```
Comment (rullzer) [inline apps/files_sharing/api/sharees.php]: Do we also want to add the other supported link headers? (see https://developer.github.com/v3/#pagination) I'm not sure if it is needed but might make sense.

File: apps/files_sharing/api/sharees.php [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		//Pagination
+		$start = ($page - 1) * $perPage;
+		$end = $page * $perPage;
+		$total = sizeof($sharees);
+
+		$sharees = array_slice($sharees, $start, $perPage);
+
+		$response = new \OC_OCS_Result($sharees);
+		$response->setTotalItems($total);
+		$response->setItemsPerPage($perPage);
```
Comment (rullzer) [inline apps/files_sharing/api/sharees.php]: Yeah it seems we have a similar issue here as with the headers in PR in #18184 since this is not reflected in the output.

Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: Well the code is being used correctly, we are not going to fix OCS in this PR :speak_no_evil:

File: apps/files_sharing/api/sharees.php (hunk 2 above)
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: The question is, if anyone will ever use them, but :speak_no_evil:

File: apps/files_sharing/api/sharees.php (hunk 1 above)
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: right, I just forgot

File: apps/files_sharing/api/sharees.php (hunk 3 above)
Comment (rullzer) [inline apps/files_sharing/api/sharees.php]: O sure...  but it is worth mentioning :)

File: apps/files_sharing/api/sharees.php (hunk 2 above)
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: updated

File: apps/files_sharing/api/sharees.php [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+	}
+
+	/**
+	 * @param string $search
+	 * @param string $sharee
+	 * @param string $label
+	 * @return bool True if the entry is an autocomplete hint, false otherwise
+	 */
+	protected function filterAutocompletion($search, $sharee, $label) {
+		if ($this->config->getSystemValue('webui-sharing-autocompletion.enabled', true)) {
```
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: https://github.com/owncloud/core/pull/18341#discussion_r37176200

File: apps/files_sharing/api/sharees.php [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+			return false;
+		}
+
+		return $search !== $sharee && $search !== $label;
```
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: https://github.com/owncloud/core/pull/18341#discussion_r37176684

File: apps/files_sharing/api/sharees.php [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+		if ($shareWithGroupOnly) {
+			// Search in all the groups this user is part of
+			$userGroups = $this->groupManager->getUserGroupIds($this->userSession->getUser());
+			foreach ($userGroups as $userGroup) {
+				$users = array_merge($users, $this->groupManager->displayNamesInGroup($userGroup, $search));
+			}
+			$users = array_unique($users);
+		} else {
+			// Search in all users
+			$users_tmp = $this->userManager->searchDisplayName($search);
```
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: Just saying, this is a performance hell, we grab all users and trim the sorted result later.
The problem is our custom sorter does not work on the user/group backend as they can only sort by uid/gid

If user enumeration is disabled, we can change this to a simple check if the group/user exists, but if type hinting is enabled it's :disappointed: 

The custom sorter moves results to the beginning of the list, which start with the entered search term. If `searchDisplayName()` would support wildcards, we could easily circumvent this, by running two requests:
1. search* and 2. *search* - both with a limit

But for now this seems to be implossible @DeepDiver1975

**Comment by DeepDiver1975** [comment]
PostgreSQL
````
OCA\Files_Sharing\Tests\API\ShareesTest::testGetShareesForShareIds with data set #1 (array(1, 2, 3), array(array('user1'), array('group1')))
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'INSERT INTO "oc_share" ("share_type", "share_with", "uid_owner", "parent", "item_type", "item_source", "item_target", "file_source", "file_target", "permissions", "stime", "accepted", "expiration", "token", "mail_send") VALUES(?, ?, ?, ?, 'fake', '', '', '0', '', '0', '0', '0', '', '', '0')' with params [0, "user1", "test7v2g8jX2wj0x5", null]:

SQLSTATE[22007]: Invalid datetime format: 7 ERROR:  invalid input syntax for type timestamp: ""
LINE 1: ..., $3, $4, 'fake', '', '', '0', '', '0', '0', '0', '', '', '0...
                                                             ^

/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php:91
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:116
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:996
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/db/connection.php:203
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:208
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/db/querybuilder/querybuilder.php:121
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/api/shareestest.php:860
/var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/api/shareestest.php:804

**Comment by nickvergessen** [comment]
hmpf, forgot to push...

Anyway, rebased and pushed now

File: apps/files_sharing/api/sharees.php (hunk 6 above)
Comment (rullzer) [inline apps/files_sharing/api/sharees.php]: I share your concern. This approach can really hurt on huge installations.
But if we want to present merged results like this I do not really see an other option.

Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: Well the old version (ajax.php) did not grab all users. The only disadvantage was, that the sorting was different (uid, gid, instead of starting with searchterm, otherwise sort by label).
I guess there are people which prefer the "nicer" way of sorting.

But I'd say it's not worth the performance troubles. This will make the drop down unusable on installations with a few hundred users/groups.

Comment (DeepDiver1975) [inline apps/files_sharing/api/sharees.php]: Perf hell: no thanks!

Comment (DeepDiver1975) [inline apps/files_sharing/api/sharees.php]: But lets move this in to allow Clients to implement this feature.
We can work in perf afterwards.

**Comment by nickvergessen** [comment]
Fixed all the points that would result in performance issues or incorrect results

**Comment by rullzer** [comment]
Why the exact part in the response? Clients know what they requested right. So they can pretty easily find the exact match on their own...

**Comment by nickvergessen** [comment]
exact parts are full matches `=== term` instead of wildcard matches `%term%`
Those should be listed first

**Comment by rullzer** [comment]
looking good :+1:

**Comment by schiesbn** [comment]
code looks good :+1:
